syntax = "proto3";

package optur;


// A value of a parameter.
message ParameterValue {
    oneof value {
        int64 int_value = 1;
        double double_value = 2;
        string string_value = 3;
    }
}

message Distribution {
    message FloatDistribution {
        float low = 1;
        float high = 2;
        bool log_scale = 3;
    }

    message IntDistribution {
        int64 low = 1;
        int64 high = 2;
        bool log_scale = 3;
    }

    message CategoricalDistribution {
        repeated ParameterValue choices = 1;
    }

    message FixedDistribution {
        ParameterValue value = 1;
    }

    message UnknownDistribution {
        repeated ParameterValue values = 1;
    }

    oneof distribution {
        UnknownDistribution unknown_distribution = 1;
        FloatDistribution float_distribution = 2;
        IntDistribution int_distribution = 3;
        CategoricalDistribution categorical_distribution = 4;
        FixedDistribution fixed_distribution = 5;
    }
    // TODO(tsuzuku): Support prior distribution.
}

// Serialized search space.
message SearchSpace {
    // A mapping from parameter-name to distribution.
    map<string, Distribution> distributions = 1;
    // TODO(tsuzuku): Support conditional search spaces.
}


message Observation {
    map<string, ParameterValue> parameters = 1;
}
